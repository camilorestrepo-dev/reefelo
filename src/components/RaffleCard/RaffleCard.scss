@import "../../styles/shared.scss";

.raffle-card-container {
  border: 1px solid $grey;
  border-radius: $radius-10;
  padding: $spacing-8;
  max-width: 300px;
  height: 150px;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: flex-start;
  cursor: pointer;

  .top-side {
    width: 100%;

    h4 {
      font-weight: bold;
      margin-top: 0;
      margin-bottom: $spacing-4;
      text-overflow: ellipsis;
      text-wrap: nowrap;
      overflow: hidden;
    }

    p {
      margin: 0;
      text-overflow: ellipsis;
      text-wrap: nowrap;
      overflow: hidden;
    }
  }

  .bottom-side {
    .info-row {
      display: flex;
      align-items: center;
      gap: $spacing-8;
      font-size: small;
    }
  }
}