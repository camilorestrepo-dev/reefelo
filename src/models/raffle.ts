export interface RaffleData {
  name: string;
  description: string;
  price?: number;
  prize: string;
  // date:
  ticketsNumber: number;
}
