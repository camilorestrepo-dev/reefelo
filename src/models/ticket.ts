export interface TicketData {
  id: string;
  reserved: boolean;
  payed: boolean;
  clientName?: string;
  clientPhone?: string;
}
