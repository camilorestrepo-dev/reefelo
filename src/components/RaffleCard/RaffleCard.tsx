import "./RaffleCard.scss";
import { RaffleData } from "../../models/raffle";
import classnames from "classnames";
import Icon from "../Icon/Icon";

type RaffleCardProps = RaffleData & {
  className?: string;
  onClick?: VoidFunction;
};

const RaffleCard: React.FC<RaffleCardProps> = ({
  className,
  description,
  name,
  ticketsNumber,
  price,
  prize,
  onClick,
}: RaffleCardProps) => {
  return (
    <div className={classnames("raffle-card-container", className)} onClick={onClick}>
      <div className="top-side">
        <h4>{name}</h4>
        <p>{description}</p>
      </div>
      <div className="bottom-side">
        <div className="info-row">
          <Icon icon="ticket" size={20} />
          {ticketsNumber}
        </div>
        <div className="info-row">
          <Icon icon="price" size={20} />${price}
        </div>
        <div className="info-row">
          <Icon icon="prize" size={20} />
          <p>{prize}</p>
        </div>
      </div>
    </div>
  );
};

export default RaffleCard;
