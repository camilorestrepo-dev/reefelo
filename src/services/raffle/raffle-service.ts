import {
  addDoc,
  collection,
  doc,
  getDocs,
  limit,
  orderBy,
  query,
  where,
  writeBatch,
  serverTimestamp,
  Timestamp,
} from "firebase/firestore";
import { RaffleData } from "../../models/raffle";
import { db } from "../firebase";
import { RAFFLES, TICKETS } from "../collections";
import { AuthService } from "../auth/auth-service";

export class RaffleService {
  static save = async (raffle: RaffleData) => {
    let tickets = raffle.tickets;
    if (!tickets) {
      const digits = Math.log10(raffle.ticketsNumber);
      tickets = Array(raffle.ticketsNumber)
        .fill(null)
        .map((_, index) => ({
          id: String(index).padStart(digits, "0"),
          enabled: true,
          payed: false,
        }));
    }
    const newRaffle = {
      ...raffle,
      userId: AuthService.currentUser().uid,
      createdAt: serverTimestamp(),
    };
    const createdDoc = await addDoc(collection(db, RAFFLES), newRaffle);
    newRaffle.id = createdDoc.id;

    const batch = writeBatch(db);
    for (const ticket of tickets) {
      batch.set(doc(db, RAFFLES, newRaffle.id, TICKETS, ticket.id), ticket);
    }
    await batch.commit();

    return newRaffle;
  };

  static loadRaffles = async (limited: number): Promise<RaffleData[]> => {
    const q = query(
      collection(db, RAFFLES),
      where("userId", "==", AuthService.currentUser().uid),
      orderBy("createdAt", "desc"),
      limit(limited),
    );
    const querySnaps = await getDocs(q);
    const raffles: RaffleData[] = [];
    querySnaps.forEach((snap) => {
      const data = snap.data();
      const raffle: RaffleData = {
        name: data.name,
        description: data.description,
        price: data.price,
        prize: data.prize,
        ticketsNumber: data.ticketsNumber,
        id: snap.id,
        userId: data.userId,
        createdAt: (data.createdAt as Timestamp).toDate(),
      };
      raffles.push(raffle);
    });
    return raffles;
  };
}
