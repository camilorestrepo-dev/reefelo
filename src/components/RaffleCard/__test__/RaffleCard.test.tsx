import { render } from "@testing-library/react";
import { RaffleData } from "../../../models/raffle";
import RaffleCard from "../RaffleCard";

const raffle: RaffleData = {
  name: "Rifa 2Motos",
  description: "Juega 25/07/25, Lote Caldas",
  price: 500,
  prize: "1000",
  ticketsNumber: 100,
};

test("reder corectly", () => {
  const component = render(<RaffleCard {...raffle} />);
  expect(component.container).toMatchSnapshot();
});
