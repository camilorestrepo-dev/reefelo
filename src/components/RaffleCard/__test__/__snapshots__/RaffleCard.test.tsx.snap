// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reder corectly 1`] = `
<div>
  <div
    class="raffle-card-container"
  >
    <div
      class="top-side"
    >
      <h4>
        Rifa 2Motos
      </h4>
      <p>
        Juega 25/07/25, Lote Caldas
      </p>
    </div>
    <div
      class="bottom-side"
    >
      <div
        class="info-row"
      >
        <img
          class="icon-ticket"
          height="20"
          src="function SvgTicket(props) {
  return React.createElement(
    'svg', 
    Object.assign({}, props, {'data-file-name': SvgTicket.name})
  );
}"
          width="20"
        />
        100
      </div>
      <div
        class="info-row"
      >
        <img
          class="icon-price"
          height="20"
          src="function SvgPrice(props) {
  return React.createElement(
    'svg', 
    Object.assign({}, props, {'data-file-name': SvgPrice.name})
  );
}"
          width="20"
        />
        $
        500
      </div>
      <div
        class="info-row"
      >
        <img
          class="icon-prize"
          height="20"
          src="function SvgPrize(props) {
  return React.createElement(
    'svg', 
    Object.assign({}, props, {'data-file-name': SvgPrize.name})
  );
}"
          width="20"
        />
        <p>
          1000
        </p>
      </div>
    </div>
  </div>
</div>
`;
